lib/getWorkflow.js

```diff
@@ -166,15 +275,36 @@ const calculateEdges = async (jobs, triggerFactory, externalDownstreamOrs, exter
                 const isJoin = upstreamAnd.size > 1;
 
                 upstreamOr.forEach(src => {
-                    edges.push({ src: filterNodeName(src), dest: jobName });
+                    if (stageObj) {
+                        const stageEdge = { src: removeStageSuffix(src), dest: jobName };
+
+                        stagesDict = updateStageDict(stagesDict, stageObj.name, stageEdge);
+                    } else {
+                        edges.push({
+                            src: removeStageSuffix(src),
+                            dest: jobName
+                        });
+                    }
                 });
+
                 upstreamAnd.forEach(src => {
                     const obj = { src, dest: jobName };
 
                     if (isJoin) {
                         obj.join = true;
                     }
-                    edges.push(obj);
+
+                    // Skip adding edge when src is stage setup
+                    if (stageObj) {
+                        if (!src.endsWith(STAGE_START)) {
```

Comment:
https://github.com/screwdriver-cd/workflow-parser/pull/43#discussion_r1243970956
`AND` and `OR` sources would be mutually exclusive. We need not skip adding the edge here.
```suggestion
```